NuRadioReco/modules/channelSinewaveSubtraction.py

```diff
@@ -0,0 +1,271 @@
+
+    def run(self, event, station, det=None, peak_prominance=4.0):
+        for channel in station.iter_channels():
+            sampling_rate = channel.get_sampling_rate()
+
+            trace = channel.get_trace()
+            trace_fil = sinewave_subtraction(
+                trace, peak_prominance, sampling_rate=sampling_rate, 
+                saved_noise_freqs=self.save_filtred_freqs)
+
+            channel.set_trace(trace_fil, sampling_rate)
+
+
+def guess_amplitude(wf: np.ndarray, target_freq: float, sampling_rate: float = 3.2):
+    """
+    Estimate the amplitude of a specific harmonic in the waveform.
+
+    Paramters
+    ----------
+    wf: np.ndarray
+        Input waveform (1D array).
+    target_freq:  float
+        Target frequency (GHz) for which to estimate amplitude.
+    sampling_rate: float (default: 3.2)
+        Sampling rate of the waveform (GHz).
+
+    Returns
+    --------
+    ampl: float
+        Estimated amplitude of the target frequency.
+    """
+    if wf.size == 0:
+        raise ValueError("Input waveform is empty.")
+
+    if target_freq < 0 or target_freq > sampling_rate / 2:
+        raise ValueError("Target frequency is out of range (0 to Nyquist frequency).")
+
+    fft_spectrum = fft.time2freq(wf, sampling_rate) * sampling_rate / np.sqrt(2) # take into account nuradio normalization
+    frequencies = fft.freqs(len(wf), sampling_rate)
+
+    # Find amplitude of the 50 Hz harmonic
+
+    bin_index = np.argmin(np.abs(frequencies - target_freq))
+    amplitude = np.abs(fft_spectrum[bin_index]) * 2 / len(wf) # Extract the amplitude (real + imaginary parts)
+
+    return amplitude
+
+
+def sinewave_subtraction(wf: np.ndarray, peak_prominance: float = 4.0, sampling_rate: float = 3.2,  saved_noise_freqs: list = None):
+    """
+    Perform sine subtraction on a waveform to remove CW noise.
+
+    Parameters
+    ----------
+    wf: np.ndarray
+        Input waveform (1D array).
+    sampling_rate: float (default: 3.2)
+        Sampling rate of the waveform (GHz).
+    peak_prominance: float (default: 6.0)
+        Threshold for identifying prominent peaks in the FFT spectrum.
+    saved_noise_freqs: list (default: None)
+        A list to store identified noise frequencies for each channel.
+    verbose: bool (default: False)
+        Print additional information.
+
+    Returns
+    -------
+    np.ndarray
+        Corrected waveform with CW noise removed.
+    """
+
+    dt = 1 / sampling_rate # in ns
+    t = np.arange(0, len(wf) * dt, dt) # in ns
+
+    # zero meaning, just in case
+    wf = wf - np.mean(wf)
+
+    def sinusoid(t, amplitude, noise_frequency, phase):
+        return amplitude * np.sin(2 * np.pi * noise_frequency * t + phase + np.pi/2)
+
+    spec = abs(fft.time2freq(wf, sampling_rate))
+    freqs = fft.freqs(len(wf), sampling_rate)
+
+    # find noise frequencies:
+    rms = np.sqrt(np.mean(np.abs(spec) ** 2))
+    peak_idxs = np.where(np.abs(spec) > peak_prominance * rms)[0]
```

Comment:
how are you handling non-CW but still relatively narrow-band signals? For example, especially for quickly-falling impulses at the edge of the low-pass, you would get a number of bins significantly larger than the rest, but this is not CW.

Reply:
ok, I see below that you're grouping them and taking the mean. I guess this is fine since most likely the iteration will be unsuccessful.